portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/clay/ClayAlert.testcase

```diff
@@ -0,0 +1,58 @@
+@component-name = "portal-frontend-infrastructure"
+definition {
+
+	property portal.release = "true";
+	property portal.upstream = "true";
+	property testray.component.names = "Clay";
+	property testray.main.component.name = "User Interface";
+
+	setUp {
+		TestCase.setUpPortalInstance();
+
+		User.firstLoginPG();
+	}
+
+	tearDown {
+		var testPortalInstance = PropsUtil.get("test.portal.instance");
+
+		if ("${testPortalInstance}" == "true") {
+			PortalInstances.tearDownCP();
+		}
+	}
+
+	@priority = "4"
+	test ToastMessageAutoClose {
+		var portalURL = PropsUtil.get("portal.url");
+
+		Navigator.openSpecificURL(url = "${portalURL}/group/control_panel/manage?p_p_id=com_liferay_configuration_admin_web_portlet_SystemSettingsPortlet&p_p_lifecycle=0&p_p_state=maximized&p_p_mode=view&_com_liferay_configuration_admin_web_portlet_SystemSettingsPortlet_mvcRenderCommandName=%2Fedit_configuration&_com_liferay_configuration_admin_web_portlet_SystemSettingsPortlet_factoryPid=com.liferay.frontend.js.spa.web.internal.configuration.SPAConfiguration&_com_liferay_configuration_admin_web_portlet_SystemSettingsPortlet_pid=com.liferay.frontend.js.spa.web.internal.configuration.SPAConfiguration");
+
+		if (IsElementPresent(locator1 = "Button#SAVE")) {
+			PortletEntry.save();
+		}
+		else {
+			Button.clickUpdate();
+		}
+
+		VerifyElementPresent(locator1 = "Message#SUCCESS_DISMISSIBLE");
```

Comment:
Hey @john-co, I don't think this should be the way to go.

If I understand the ticket correctly, we simply want to assert that toasts show and can be dismissed or go away after a certain time.

A lot can go wrong here that is totally unrelated to how toasts work and that couple the test with huge portions of the platform:
- What happens if there are problems saving configurations?
- What if we remove the configuration?
- What if those URLs change?

As a rule of thumb, we should isolate our tests as much as possible from other parts of the system. This is one of the main causes of frustration when trying to change things in portal. Completely unrelated tests using adjacent features in a completely unexpected way.

You should explore what makes the feature tick, get to its core and then test for it. Otherwise, we're just looking for side effects and that's very unreliable.

Some ideas off the top of my head:
- To validate that success messages appear on successful actions: 
    - Create a test action command that adds a success message to the request and redirects to the same page. 
- To validate that the JS APIs generate the proper messages: 
    - Create a JS snippet or a dummy widget with your desired functionality to call the `Liferay.Util.openToast` in a JSP
    - Check for existence of unit tests of `openToast`. If they exist, validate it tests the expectations. If they don't, create them or ask for them to be created 

/cc @PablitoBonito, @zhangyan0701, @manoelcyreno, please let me know what you think of this. I'm open to work on this together if you need it... maybe we could start by reviewing the test cases and get them through our workflow just as we do with the rest of the team. 
/cc2 @wincent, @dsanz, in case you want to add to the conversation

Reply:
I'm not sure if we've considered this in the past: what if we leverage [frontend-taglib-clay-sample-web](https://github.com/liferay/liferay-portal/tree/master/modules/apps/frontend-taglib/frontend-taglib-clay-sample-web) to make it the target of tests like this?. 

Basically, tests would be very similar to the one here, but rather than looking for portal usages of our tags, test would add some of the sample portlets to a page and test there. You may consider this a "trick" beause we control the sample portlet, but this is happening in unit tests around the portal. Perhaps the anomaly is to find places in the portal where our tag is used and rely on such usage to test.

Besides showcasing the API, it would allow us to isolate the tests by providing the right elements, and no more, where the thing being tested is used as expected. A change in the taglib API would immediately reflect in the corresponding sample-web portlet and the test, w/o having to wait until someone uses it, find the usage, etc.

I don't have a full vision of what other sample -web bundles would be required to fulfill this vision, we may start this in the context of clay taglib samples and see if it makes sense for other visual components.

As a side effect, that would set clear expectations on tested usage patterns for the frontend devs at product teams, sort of "here are our components, how to use them and the test asserting the usage works as expected"

Reply:
Yeah sure, the core of the proposal is not necessarily reuse these sample-web projects but identify what could be built in order to get closer to the functionality units we want to test, and no more. Perhaps these test apps will evolve faster than the sample-web projects to keep up with the new features.

I'm proposing this approach in the hope that it's better than finding specific usages of the infra in the portal, then test those. With "better" I mean that the "specific usages" of the infra in the products shall hopefully be covered by the tests specifically made for the features in question, not from the infra tests themselves. This leaves room for more targeted, lightweight, reliable and maintainable tests. 

I understand there might be opposed opinions on this because what adds value are the "real" usages of the infra and not the ones we create via "synthetic" elements. Furthermore, considering that part of our "features" are related to platform improvements and tech debt, it's even harder to find what to test, how can it be observed and right behavior asserted (as compared to a "end-user feature which directly reflects in the UI/tooling". I don't know how are we designing the testing so I can't tell how much of this approach can be implemented successfully.

Reply:
> I'm proposing this approach in the hope that it's better than finding specific usages of the infra in the portal, then test those. With "better" I mean that the "specific usages" of the infra in the products shall hopefully be covered by the tests specifically made for the features in question, not from the infra tests themselves. This leaves room for more targeted, lightweight, reliable and maintainable tests.

Totally agree on this, I think we're definitely on the same page. It's up to @john-co, @manoelcyreno, @PablitoBonito and @zhangyan0701 to weigh in. I'd be happy to help steer our testing efforts in that direction.

I'd earnestly ask that we start at least getting QA PRs through our regular workflow and following our regular guidelines so we can all better collaborate on this.

Reply:
Hi @jbalsas and @dsanz, thank you for your feedback on this. I agree and am on board with the approach you're both getting at.

I added this test as a temporary coverage patch up and was planning to follow up later with a more isolated test. I was planning to use the [JS Clay Sample App](https://github.com/john-co/liferay-portal/pull/324) @bryceosterhaus had proposed a while back but wanted to get some sort of coverage in first before diving into the app. Is this still in the direction we're looking for? Either way, I would definitely appreciate more coaching on creating the dummy portlets and making the API calls because I have not done them before. I'll also look more into [the links](https://liferay.slack.com/archives/G3JBR21HA/p1602594027423300) Manoel recommended since they were not a part of my liferay training/onboarding.

I agree with sending the QA PRs through the regular workflow so we can better collaborate and be on the same page.

Reply:
> I was planning to use the JS Clay Sample App @bryceosterhaus had proposed a while back

We would be in the same position. A Sample App is a sample app and has a purpose. In this case, @bryceosterhaus built it so people could easily test snippets of code or make quick tests, not as a gateway for all the tests frontend.

As a rule of thumb, we should **never** use other people's (even this team) developments for our tests. We should **build** our own test widgets or other artifacts that we can control and link to the test lifecycle.

Reply:
Hi @jbalsas, 

ok, we won't use the js clay sample app for automation testing. What is the plan for building the widgets for automation? Who will build them?

Reply:
> What is the plan for building the widgets for automation? Who will build them?

Ideally it should be whoever is building the test. That's the only way to make sure the artefact matches the test requirements and are kept in sync or updated accordingly to test changes. It's also the only way that those can be maintained without creating bottlenecks or unnecessary coupling between team members.

That being said, if whoever is building the test does not have the skills to do it, someone else in the team could build it. Then we should make sure the knowledge is transferred and future maintenance can be owned so we don't need to ask for help for every little adjustment in the testing widget.